### ISSUE TITLE
Ensemble race: only root starts up

### ISSUE THREAD

**Issue by engelsanchez** [basho/riak_kv#983]
While testing https://github.com/basho/riak_kv/pull/980 I got a cluster into a state where ensembles other than root will never start.  Even though the root ensemble lists one ensemble per preflist as expected (the return value of riak_ensemble_manager:known_ensembles()), the manager's cluster state only contains an entry for the root ensemble.  So the ensembles will never be started.  The output of riak-admin ensemble-status is stuck on:

```
$ dev3/bin/riak-admin ensemble-status
============================== Consensus System ===============================
Enabled:     true
Active:      true
Ring Ready:  true
Paranoia:    low (syncing not required)
AAE enabled: true

================================== Ensembles ==================================
 Ensemble     Quorum        Nodes      Leader
-------------------------------------------------------------------------------
   root       1 / 1         1 / 1      dev1@127.0.0.1
```

My 3 node cluster got into that state after a quick restart following a change to the paranoia config value from medium to low.  I still have that cluster up in that state.  I set up another 3 node cluster the same way and so far it has not hit the same problem.  I will add more details if I figure out anything after some more debugging.  I am not sure if this is a KV or ensemble issue. This issue will be moved to riak_ensemble if data points that way.

**Comment by jtuple** [comment]
I'd love to see see the logs for this cluster if you still have them.

**Comment by engelsanchez** [comment]
The logs are here: https://s3.amazonaws.com/uploads.hipchat.com/20796/189387/Q7EnwMf4svXcLsf/ensemble-stuck-logs.tgz

Basically the root ensemble somehow has a different cluster version. It tries to gossip the entire ensemble list to everybody, but it is rejected and you get a lot of this in the logs:

```
2014-06-16 09:24:00.847 [warning] <0.268.0>@riak_ensemble_state:merge:173 Ignoring cluster state with different id
```

**Comment by jtuple** [comment]
Different cluster version? Interesting, that should only occur if `riak_ensemble_manager:enable()` was called on more than one node. Unless you were manually calling `enable()` yourself (please don't), that shouldn't happen unless there's a bug in the previously merged basho/riak_core#601. Such fun! More things to investigate.

**Comment by engelsanchez** [comment]
I didn't call it manually. I've been letting the cluster do its thing. It is possible that this happened because the claimant started with a deleted data directory and somehow created and activated on the initial state before the root ensemble was ready. I started 3 nodes simultaneously with a little script. That must somehow contributed, but by itself has not been enough to reproduce. That's just a guess and I don't remember deleting it. I'm trying to understand the bootstrapping code path a bit better and will try a couple of nasty things I might have done.

**Comment by jtuple** [comment]
FYI, deleting the `data/ring` directory but not the `data/ensembles` directory should be one way to reproduce, but you're basically reproducing a "don't do that!" scenario.

**Comment by engelsanchez** [comment]
Adding this info I forgot to mention: The version (which is just {node, time}) of both cluster states show the same node: 'dev1@127.0.0.1', the claimant in my setup. The time of the initial state with root only is 3 minutes ahead of the time of the good, complete cluster state:

```
1> calendar:now_to_local_time({1402,684037,539719}).
{{2014,6,13},{14,27,17}}
2> calendar:now_to_local_time({1402,683855,48463}).
{{2014,6,13},{14,24,15}}
```

**Comment by engelsanchez** [comment]
@jtuple would instructions on recovering a failed node need to be updated because of the issue you mentioned above? They tell users to delete their ring directories and re-join the cluster. http://docs.basho.com/riak/latest/ops/running/recovery/failed-node/

**Comment by krestenkrab** [comment]
It would seem that a missing `data/ring` directory, but at the same time an existing `data/ensembles` direction is a simple condition to test/warn about.

**Comment by engelsanchez** [comment]
We believe this issue may have something to do with the fact that the root ensemble state is written quickly, whereas the manager and facts are periodically sync'ed to disk.  So the first write takes a bit to actually making it to disk. During that time, an unlucky combination of node stops and restarts could end up in the situation described here. For that, issue basho/riak_ensemble#34 has been opened.

**Comment by engelsanchez** [comment]
Ok. I've managed to reproduce this issue. This is the timeline of events:
- Start and join 3 nodes into a cluster with strong consistency on
- The claimant activates the cluster which as you can see [here](https://github.com/basho/riak_ensemble/blob/f7807bee9dfc259ad6b9c1c7c6fd6a982658e27c/src/riak_ensemble_manager.erl#L296) didn't use to save the state right away. It would be saved on the next storage tick, which could take seconds (see [here](https://github.com/basho/riak_ensemble/blob/f7807bee9dfc259ad6b9c1c7c6fd6a982658e27c/src/riak_ensemble_storage.erl#L67) and [here](https://github.com/basho/riak_ensemble/blob/f7807bee9dfc259ad6b9c1c7c6fd6a982658e27c/src/riak_ensemble_storage.erl#L172).
- Quickly stop all nodes before the next storage tick after ensemble-status lists ensembles for all preflists.  This won't happen quickly enough unless trust is high and AAE is not required to start them
- The ensemble facts files are missing! But the root ensemble data is there (files ending in _kv in data/ensembles).  The root ensemble stores all the ensembles already created in that file.
- Start again, with no facts file, the claimant decides to activate again. (see [here](https://github.com/basho/riak_core/blob/develop/src/riak_core_claimant.erl#L756)).
- The root ensemble will gossip the cluster state with the ensemble list to everybody, which would fix everything. But there is a chance this will happen after the claimant creates a new cluster state above. I made the manager tick configurable to make sure the manager wins and broadcasts the newly created state first
- Bingo! Now the manager thinks there should only be a root ensemble.  The code that would populate the ensemble list based on the ring is [here](https://github.com/basho/riak_kv/blob/4141ff1f8dd61bdc8a6c9ed203394dbedcdd38d3/src/riak_kv_ensembles.erl#L171).  But that calls [riak_ensemble_manager:known_ensembles](https://github.com/basho/riak_ensemble/blob/f7807bee9dfc259ad6b9c1c7c6fd6a982658e27c/src/riak_ensemble_manager.erl#L170) to figure out what ensembles already exists, but that queries the root ensemble directly, so it sees all the entire original list and thinks that no new ones need to be created.
- When the root ensemble gossips its old cluster state, nobody listens because it has a different version than the newly created one. (see [riak_ensemble_state:merge](https://github.com/basho/riak_ensemble/blob/f7807bee9dfc259ad6b9c1c7c6fd6a982658e27c/src/riak_ensemble_state.erl#L172)).  And we are stuck forever

With the fix in basho/riak_ensemble#35 though, the manager state is saved immediately and disaster is avoided.

That's a long way to say I'm closing this issue.

**Comment by engelsanchez** [comment]
And for the record, I think that perhaps making the root ensemble use the ensemble storage mechanism so that all facts and root ensemble data are saved atomically might prevent other scenarios where the data diverges due to a series of unfortunate events.
